File: Programmers/KAKAO/2020/괄호변환.py
def division(p):
    cnt = [0, 0]
    for i in range(len(p)):
        if p[i] == '(':
            cnt[0] += 1
        else:
            cnt[1] += 1
        if cnt[0] == cnt[1]:
            return (p[:i+1], p[i+1:])
        

def isRight(p):
    stack = list()
    for s in p:
        if s == '(':
            stack.append(p)
        else:
            if stack:
                stack.pop()
            else:
                return False
    return True

def solution(p):

    if not p:
        return ''
    
    u, v = division(p)

    if isRight(u):
        return u + solution(v)
    else:
        answer = '('
        answer += solution(v)
        answer += ')'

        for i in range(1, len(u)-1):
            if u[i] == '(':
                answer += ')'
            else:
                answer += '('
        
        return answer

p = "()))((()"
print(solution(p))

File: Programmers/Level2/괄호변환/괄호변환.js
function division(p) {
  const cnt = [0, 0];
  for (let i = 0; i < p.length; i++) {
    if (p[i] === "(") {
      cnt[0] += 1;
    } else {
      cnt[1] += 1;
    }
    if (cnt[0] == cnt[1]) return [p.slice(0, i + 1), p.slice(i + 1)];
  }
}

function isRight(p) {
  const stack = [];
  for (let i = 0; i < p.length; i++) {
    if (p[i] === "(") {
      stack.push(p[i]);
    } else {
      if (stack.length) {
        stack.pop();
      } else {
        return false;
      }
    }
  }
  return true;
}

function solution(p) {
  if (!p.length) return "";

  const [u, v] = division(p);

  if (isRight(u)) {
    return u + solution(v);
  } else {
    let answer = "(";
    answer += solution(v);
    answer += ")";

    for (let i = 1; i < u.length - 1; i++) {
      if (u[i] === "(") {
        answer += ")";
      } else {
        answer += "(";
      }
    }
    return answer;
  }
}

const p = "(()())()";
console.log(solution(p));
